File: Kotlin/json_utils/json_utils_test.kt
package json_utils

/**
 * JSON Utilities Test Suite
 * Tests all JSON parsing and generation functionality
 */

fun main() {
    println("Running JSON Utils Tests...")
    println("=" * 50)

    var passed = 0
    var failed = 0

    // Test 1: Parse null
    try {
        val result = JsonUtils.parse("null")
        assert(result.isNull()) { "Expected null" }
        println("✓ Test 1: Parse null")
        passed++
    } catch (e: Exception) {
        println("✗ Test 1: Parse null - ${e.message}")
        failed++
    }

    // Test 2: Parse boolean true
    try {
        val result = JsonUtils.parse("true")
        assert(result.isBoolean() && result.asBoolean()) { "Expected true" }
        println("✓ Test 2: Parse boolean true")
        passed++
    } catch (e: Exception) {
        println("✗ Test 2: Parse boolean true - ${e.message}")
        failed++
    }

    // Test 3: Parse boolean false
    try {
        val result = JsonUtils.parse("false")
        assert(result.isBoolean() && !result.asBoolean()) { "Expected false" }
        println("✓ Test 3: Parse boolean false")
        passed++
    } catch (e: Exception) {
        println("✗ Test 3: Parse boolean false - ${e.message}")
        failed++
    }

    // Test 4: Parse integer
    try {
        val result = JsonUtils.parse("42")
        assert(result.isNumber() && result.asInt() == 42) { "Expected 42" }
        println("✓ Test 4: Parse integer")
        passed++
    } catch (e: Exception) {
        println("✗ Test 4: Parse integer - ${e.message}")
        failed++
    }

    // Test 5: Parse negative number
    try {
        val result = JsonUtils.parse("-123")
        assert(result.isNumber() && result.asInt() == -123) { "Expected -123" }
        println("✓ Test 5: Parse negative number")
        passed++
    } catch (e: Exception) {
        println("✗ Test 5: Parse negative number - ${e.message}")
        failed++
    }

    // Test 6: Parse float
    try {
        val result = JsonUtils.parse("3.14")
        assert(result.isNumber() && kotlin.math.abs(result.asDouble() - 3.14) < 0.001) { "Expected 3.14" }
        println("✓ Test 6: Parse float")
        passed++
    } catch (e: Exception) {
        println("✗ Test 6: Parse float - ${e.message}")
        failed++
    }

    // Test 7: Parse scientific notation
    try {
        val result = JsonUtils.parse("1.5e10")
        assert(result.isNumber() && result.asDouble() == 1.5e10) { "Expected 1.5e10" }
        println("✓ Test 7: Parse scientific notation")
        passed++
    } catch (e: Exception) {
        println("✗ Test 7: Parse scientific notation - ${e.message}")
        failed++
    }

    // Test 8: Parse simple string
    try {
        val result = JsonUtils.parse("\"hello\"")
        assert(result.isString() && result.asString() == "hello") { "Expected hello" }
        println("✓ Test 8: Parse simple string")
        passed++
    } catch (e: Exception) {
        println("✗ Test 8: Parse simple string - ${e.message}")
        failed++
    }

    // Test 9: Parse string with escape sequences
    try {
        val result = JsonUtils.parse("\"hello\\nworld\"")
        assert(result.isString() && result.asString() == "hello\nworld") { "Expected hello\\nworld" }
        println("✓ Test 9: Parse string with escape sequences")
        passed++
    } catch (e: Exception) {
        println("✗ Test 9: Parse string with escape sequences - ${e.message}")
        failed++
    }

    // Test 10: Parse empty array
    try {
        val result = JsonUtils.parse("[]")
        assert(result.isArray() && result.size() == 0) { "Expected empty array" }
        println("✓ Test 10: Parse empty array")
        passed++
    } catch (e: Exception) {
        println("✗ Test 10: Parse empty array - ${e.message}")
        failed++
    }

    // Test 11: Parse array with values
    try {
        val result = JsonUtils.parse("[1, 2, 3]")
        assert(result.isArray() && result.size() == 3) { "Expected array with 3 elements" }
        val arr = result.asArray()
        assert(arr.getInt(0) == 1 && arr.getInt(1) == 2 && arr.getInt(2) == 3)
        println("✓ Test 11: Parse array with values")
        passed++
    } catch (e: Exception) {
        println("✗ Test 11: Parse array with values - ${e.message}")
        failed++
    }

    // Test 12: Parse nested array
    try {
        val result = JsonUtils.parse("[[1, 2], [3, 4]]")
        assert(result.isArray() && result.size() == 2) { "Expected nested array" }
        val arr = result.asArray()
        assert(arr.getArray(0).size() == 2 && arr.getArray(1).size() == 2)
        println("✓ Test 12: Parse nested array")
        passed++
    } catch (e: Exception) {
        println("✗ Test 12: Parse nested array - ${e.message}")
        failed++
    }

    // Test 13: Parse empty object
    try {
        val result = JsonUtils.parse("{}")
        assert(result.isObject() && result.size() == 0) { "Expected empty object" }
        println("✓ Test 13: Parse empty object")
        passed++
    } catch (e: Exception) {
        println("✗ Test 13: Parse empty object - ${e.message}")
        failed++
    }

    // Test 14: Parse simple object
    try {
        val result = JsonUtils.parse("""{"name": "John", "age": 30}""")
        assert(result.isObject() && result.size() == 2) { "Expected object with 2 properties" }
        val obj = result.asObject()
        assert(obj.getString("name") == "John" && obj.getInt("age") == 30)
        println("✓ Test 14: Parse simple object")
        passed++
    } catch (e: Exception) {
        println("✗ Test 14: Parse simple object - ${e.message}")
        failed++
    }

    // Test 15: Parse nested object
    try {
        val result = JsonUtils.parse("""{"user": {"name": "John", "id": 1}}""")
        assert(result.isObject()) { "Expected object" }
        val obj = result.asObject()
        assert(obj.has("user"))
        val user = obj.getObject("user")
        assert(user.getString("name") == "John" && user.getInt("id") == 1)
        println("✓ Test 15: Parse nested object")
        passed++
    } catch (e: Exception) {
        println("✗ Test 15: Parse nested object - ${e.message}")
        failed++
    }

    // Test 16: Test object get with default
    try {
        val result = JsonUtils.parse("""{"name": "John"}""")
        val obj = result.asObject()
        assert(obj.getString("missing", "default") == "default")
        assert(obj.getInt("missing", 42) == 42)
        assert(obj.getBoolean("missing", true))
        println("✓ Test 16: Object get with default")
        passed++
    } catch (e: Exception) {
        println("✗ Test 16: Object get with default - ${e.message}")
        failed++
    }

    // Test 17: Test array get with default
    try {
        val result = JsonUtils.parse("[1, 2]")
        val arr = result.asArray()
        assert(arr.getInt(10, 99) == 99)
        assert(arr.getString(10, "default") == "default")
        println("✓ Test 17: Array get with default")
        passed++
    } catch (e: Exception) {
        println("✗ Test 17: Array get with default - ${e.message}")
        failed++
    }

    // Test 18: Generate JSON from object
    try {
        val obj = JsonUtils.obj(
            "name" to "John",
            "age" to 30,
            "active" to true
        )
        val json = obj.toJsonString()
        assert(json.contains("\"name\":\"John\""))
        assert(json.contains("\"age\":30"))
        assert(json.contains("\"active\":true"))
        println("✓ Test 18: Generate JSON from object")
        passed++
    } catch (e: Exception) {
        println("✗ Test 18: Generate JSON from object - ${e.message}")
        failed++
    }

    // Test 19: Generate JSON from array
    try {
        val arr = JsonUtils.arr(1, 2, 3, "hello")
        val json = arr.toJsonString()
        assert(json == "[1,2,3,\"hello\"]")
        println("✓ Test 19: Generate JSON from array")
        passed++
    } catch (e: Exception) {
        println("✗ Test 19: Generate JSON from array - ${e.message}")
        failed++
    }

    // Test 20: Pretty print JSON
    try {
        val json = """{"name":"John","age":30}"""
        val pretty = JsonUtils.prettyPrint(json)
        assert(pretty.contains("{\n"))
        assert(pretty.contains("  \"name\":"))
        println("✓ Test 20: Pretty print JSON")
        passed++
    } catch (e: Exception) {
        println("✗ Test 20: Pretty print JSON - ${e.message}")
        failed++
    }

    // Test 21: Minify JSON
    try {
        val json = """{
            "name": "John",
            "age": 30
        }"""
        val minified = JsonUtils.minify(json)
        assert(!minified.contains("\n"))
        assert(!minified.contains("  "))
        println("✓ Test 21: Minify JSON")
        passed++
    } catch (e: Exception) {
        println("✗ Test 21: Minify JSON - ${e.message}")
        failed++
    }

    // Test 22: Merge objects
    try {
        val obj1 = JsonUtils.obj("a" to 1, "b" to 2)
        val obj2 = JsonUtils.obj("b" to 3, "c" to 4)
        val merged = JsonUtils.merge(obj1, obj2)
        assert(merged.getInt("a") == 1)
        assert(merged.getInt("b") == 3) // obj2 overrides
        assert(merged.getInt("c") == 4)
        println("✓ Test 22: Merge objects")
        passed++
    } catch (e: Exception) {
        println("✗ Test 22: Merge objects - ${e.message}")
        failed++
    }

    // Test 23: Validate JSON
    try {
        assert(JsonUtils.isValid("{}"))
        assert(JsonUtils.isValid("[]"))
        assert(JsonUtils.isValid("\"hello\""))
        assert(JsonUtils.isValid("123"))
        assert(JsonUtils.isValid("true"))
        assert(!JsonUtils.isValid("{invalid}"))
        assert(!JsonUtils.isValid(""))
        println("✓ Test 23: Validate JSON")
        passed++
    } catch (e: Exception) {
        println("✗ Test 23: Validate JSON - ${e.message}")
        failed++
    }

    // Test 24: Parse or null
    try {
        assert(JsonUtils.parseOrNull("{}") != null)
        assert(JsonUtils.parseOrNull("invalid") == null)
        println("✓ Test 24: Parse or null")
        passed++
    } catch (e: Exception) {
        println("✗ Test 24: Parse or null - ${e.message}")
        failed++
    }

    // Test 25: Complex nested structure
    try {
        val json = """{
            "users": [
                {"id": 1, "name": "Alice", "roles": ["admin", "user"]},
                {"id": 2, "name": "Bob", "roles": ["user"]}
            ],
            "meta": {
                "total": 2,
                "page": 1
            }
        }"""
        val result = JsonUtils.parse(json)
        assert(result.isObject())
        val users = result.asObject().getArray("users")
        assert(users.size() == 2)
        assert(users.getObject(0).getString("name") == "Alice")
        assert(users.getObject(0).getArray("roles").size() == 2)
        assert(users.getObject(1).getString("name") == "Bob")
        assert(result.asObject().getObject("meta").getInt("total") == 2)
        println("✓ Test 25: Complex nested structure")
        passed++
    } catch (e: Exception) {
        println("✗ Test 25: Complex nested structure - ${e.message}")
        failed++
    }

    // Test 26: Unicode escape sequences
    try {
        val result = JsonUtils.parse("\"\\u0048\\u0065\\u006c\\u006c\\u006f\"")
        assert(result.asString() == "Hello")
        println("✓ Test 26: Unicode escape sequences")
        passed++
    } catch (e: Exception) {
        println("✗ Test 26: Unicode escape sequences - ${e.message}")
        failed++
    }

    // Test 27: Special characters in strings
    try {
        val result = JsonUtils.parse("\"hello\\tworld\\nnew line\"")
        assert(result.asString() == "hello\tworld\nnew line")
        println("✓ Test 27: Special characters in strings")
        passed++
    } catch (e: Exception) {
        println("✗ Test 27: Special characters in strings - ${e.message}")
        failed++
    }

    // Test 28: Null values in object
    try {
        val result = JsonUtils.parse("""{"name": null, "value": 42}""")
        assert(result.asObject().get("name").isNull())
        assert(result.asObject().getInt("value") == 42)
        println("✓ Test 28: Null values in object")
        passed++
    } catch (e: Exception) {
        println("✗ Test 28: Null values in object - ${e.message}")
        failed++
    }

    // Test 29: Mixed types in array
    try {
        val result = JsonUtils.parse("[1, \"two\", true, null, {\"a\": 1}]")
        val arr = result.asArray()
        assert(arr.getInt(0) == 1)
        assert(arr.getString(1) == "two")
        assert(arr.getBoolean(2))
        assert(arr[3].isNull())
        assert(arr.getObject(4).getInt("a") == 1)
        println("✓ Test 29: Mixed types in array")
        passed++
    } catch (e: Exception) {
        println("✗ Test 29: Mixed types in array - ${e.message}")
        failed++
    }

    // Test 30: Round-trip parsing
    try {
        val original = """{"name":"John","items":[1,2,3],"active":true}"""
        val parsed = JsonUtils.parse(original)
        val generated = parsed.toJsonString()
        val reparsed = JsonUtils.parse(generated)
        assert(reparsed.asObject().getString("name") == "John")
        assert(reparsed.asObject().getArray("items").size() == 3)
        assert(reparsed.asObject().getBoolean("active"))
        println("✓ Test 30: Round-trip parsing")
        passed++
    } catch (e: Exception) {
        println("✗ Test 30: Round-trip parsing - ${e.message}")
        failed++
    }

    println("=" * 50)
    println("Results: $passed passed, $failed failed")
    if (failed == 0) {
        println("All tests passed! ✓")
    } else {
        println("Some tests failed! ✗")
        kotlin.system.exitProcess(1)
    }
}

File: MATLAB/json_utils/json_utils_test.m
function json_utils_test()
    %JSON_UTILS_TEST Test suite for JSON Utilities
    %   Runs all tests for the JSON parsing and generation utilities.
    %
    %   Usage:
    %       json_utils_test()

    fprintf('Running JSON Utilities Test Suite...\n');
    fprintf('====================================\n\n');

    utils = json_utils.mod();
    passed = 0;
    failed = 0;

    %% Test 1: Parse empty object
    try
        result = utils.parse('{}');
        assert(isstruct(result) && isempty(fieldnames(result)), 'Empty object test failed');
        fprintf('✓ Test 1: Parse empty object\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 1: Parse empty object - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 2: Parse empty array
    try
        result = utils.parse('[]');
        assert(iscell(result) && isempty(result), 'Empty array test failed');
        fprintf('✓ Test 2: Parse empty array\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 2: Parse empty array - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 3: Parse simple object
    try
        result = utils.parse('{"name": "John", "age": 30}');
        assert(strcmp(result.name, 'John') && result.age == 30, 'Simple object test failed');
        fprintf('✓ Test 3: Parse simple object\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 3: Parse simple object - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 4: Parse nested object
    try
        result = utils.parse('{"user": {"name": "Alice", "id": 123}}');
        assert(strcmp(result.user.name, 'Alice') && result.user.id == 123, 'Nested object test failed');
        fprintf('✓ Test 4: Parse nested object\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 4: Parse nested object - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 5: Parse array of numbers
    try
        result = utils.parse('[1, 2, 3, 4, 5]');
        assert(iscell(result) && length(result) == 5 && result{3} == 3, 'Number array test failed');
        fprintf('✓ Test 5: Parse array of numbers\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 5: Parse array of numbers - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 6: Parse array of strings
    try
        result = utils.parse('["apple", "banana", "cherry"]');
        assert(iscell(result) && strcmp(result{2}, 'banana'), 'String array test failed');
        fprintf('✓ Test 6: Parse array of strings\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 6: Parse array of strings - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 7: Parse mixed array
    try
        result = utils.parse('[1, "two", true, null]');
        assert(result{1} == 1 && strcmp(result{2}, 'two') && islogical(result{3}) && isempty(result{4}), ...
               'Mixed array test failed');
        fprintf('✓ Test 7: Parse mixed array\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 7: Parse mixed array - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 8: Parse true/false/null
    try
        result = utils.parse('{"active": true, "deleted": false, "data": null}');
        assert(result.active == true && result.deleted == false && isempty(result.data), ...
               'Boolean/null test failed');
        fprintf('✓ Test 8: Parse true/false/null\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 8: Parse true/false/null - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 9: Parse numbers (integer and float)
    try
        result = utils.parse('{"int": 42, "float": 3.14, "negative": -10, "exp": 1.5e10}');
        assert(result.int == 42 && abs(result.float - 3.14) < 0.001 && result.negative == -10, ...
               'Number parsing test failed');
        fprintf('✓ Test 9: Parse numbers\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 9: Parse numbers - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 10: Parse string with escapes
    try
        result = utils.parse('{"text": "Line 1\\nLine 2\\tTabbed"}');;
        assert(contains(result.text, sprintf('\n')) && contains(result.text, sprintf('\t')), ...
               'String escape test failed');
        fprintf('✓ Test 10: Parse string with escapes\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 10: Parse string with escapes - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 11: Encode simple struct
    try
        data = struct('name', 'Test', 'value', 123);
        json = utils.encode(data);
        assert(contains(json, '"name":"Test"') && contains(json, '"value":123'), ...
               'Encode struct test failed');
        fprintf('✓ Test 11: Encode simple struct\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 11: Encode simple struct - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 12: Encode cell array
    try
        data = {1, 'two', true};
        json = utils.encode(data);
        assert(strcmp(json, '[1,"two",true]'), 'Encode cell array test failed');
        fprintf('✓ Test 12: Encode cell array\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 12: Encode cell array - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 13: Encode pretty print
    try
        data = struct('a', 1, 'b', 2);
        json = utils.encodePretty(data);
        assert(contains(json, sprintf('\n')) && contains(json, '  '), ...
               'Pretty print test failed');
        fprintf('✓ Test 13: Encode pretty print\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 13: Encode pretty print - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 14: Round-trip test
    try
        original = struct('name', 'Test', 'items', {{1, 2, 3}}, 'active', true);
        json = utils.encode(original);
        result = utils.parse(json);
        assert(strcmp(result.name, 'Test') && result.active == true, ...
               'Round-trip test failed');
        fprintf('✓ Test 14: Round-trip encode/decode\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 14: Round-trip encode/decode - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 15: isValid with valid JSON
    try
        assert(utils.isValid('{"valid": true}') == true, 'isValid true test failed');
        fprintf('✓ Test 15: isValid with valid JSON\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 15: isValid with valid JSON - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 16: isValid with invalid JSON
    try
        assert(utils.isValid('invalid json') == false, 'isValid false test failed');
        fprintf('✓ Test 16: isValid with invalid JSON\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 16: isValid with invalid JSON - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 17: parseOrNull with valid JSON
    try
        result = utils.parseOrNull('{"key": "value"}');
        assert(strcmp(result.key, 'value'), 'parseOrNull valid test failed');
        fprintf('✓ Test 17: parseOrNull with valid JSON\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 17: parseOrNull with valid JSON - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 18: parseOrNull with invalid JSON
    try
        result = utils.parseOrNull('not json');
        assert(isempty(result), 'parseOrNull invalid test failed');
        fprintf('✓ Test 18: parseOrNull with invalid JSON\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 18: parseOrNull with invalid JSON - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 19: Get with default value
    try
        data = struct('existing', 'value');
        result = utils.get(data, 'existing', 'default');
        assert(strcmp(result, 'value'), 'Get existing test failed');
        result = utils.get(data, 'missing', 'default');
        assert(strcmp(result, 'default'), 'Get default test failed');
        fprintf('✓ Test 19: Get with default value\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 19: Get with default value - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 20: GetPath nested access
    try
        data = struct('level1', struct('level2', struct('value', 'deep')));
        result = utils.getPath(data, 'level1.level2.value', 'default');
        assert(strcmp(result, 'deep'), 'GetPath test failed');
        result = utils.getPath(data, 'level1.missing.value', 'default');
        assert(strcmp(result, 'default'), 'GetPath default test failed');
        fprintf('✓ Test 20: GetPath nested access\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 20: GetPath nested access - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 21: Minify JSON
    try
        minified = utils.minify('{  "a"  :  1  ,  "b"  :  2  }');
        assert(strcmp(minified, '{"a":1,"b":2}'), 'Minify test failed');
        fprintf('✓ Test 21: Minify JSON\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 21: Minify JSON - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 22: Static quickParse
    try
        result = json_utils.mod.quickParse('{"static": true}');
        assert(result.static == true, 'Static quickParse test failed');
        fprintf('✓ Test 22: Static quickParse\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 22: Static quickParse - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 23: Static quickEncode
    try
        json = json_utils.mod.quickEncode(struct('quick', 'test'));
        assert(contains(json, '"quick":"test"'), 'Static quickEncode test failed');
        fprintf('✓ Test 23: Static quickEncode\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 23: Static quickEncode - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 24: Parse array of objects
    try
        result = utils.parse('[{"id": 1}, {"id": 2}, {"id": 3}]');
        assert(iscell(result) && length(result) == 3 && result{2}.id == 2, ...
               'Array of objects test failed');
        fprintf('✓ Test 24: Parse array of objects\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 24: Parse array of objects - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Test 25: Encode empty struct
    try
        json = utils.encode(struct());
        assert(strcmp(json, '{}'), 'Encode empty struct test failed');
        fprintf('✓ Test 25: Encode empty struct\n');
        passed = passed + 1;
    catch ME
        fprintf('✗ Test 25: Encode empty struct - %s\n', ME.message);
        failed = failed + 1;
    end

    %% Summary
    fprintf('\n====================================\n');
    fprintf('Test Results: %d passed, %d failed\n', passed, failed);

    if failed == 0
        fprintf('All tests passed! ✓\n');
    else
        fprintf('Some tests failed. ✗\n');
    end
end
